julia/mmtk_julia.c

```diff
@@ -138,65 +141,89 @@ void mmtk_exit_from_safepoint(int8_t old_state) {
     jl_gc_state_set(ptls, old_state, JL_GC_STATE_WAITING);
 }
 
-// all threads pass here and if there is another thread doing GC,
-// it will block until GC is done
-// that thread simply exits from block_for_gc without executing finalizers
-// when executing finalizers do not let another thread do GC (set a variable such that while that variable is true, no GC can be done)
-int8_t set_gc_initial_state(void* ptls_raw) 
+// based on jl_gc_collect from gc.c
+JL_DLLEXPORT void jl_gc_prepare_to_collect(void)
```

Comment:
Looks like this function just duplicates `jl_gc_collect`. Have you considered moving the function to `gc-common.c`, and letting Julia and MMTk each implement a different `_jl_gc_collect`?

Reply:
The problem is that triggering GC manually in Julia is done via a call to `jl_gc_collect` (https://github.com/mmtk/julia/blob/83796a7d38878faa9afdb94ce84b6da00035b336/base/gcutils.jl#L129). For MMTk, we want that function to call into `mmtk_handle_user_collection_request` as it currently does. That's why I did the duplication, with `jl_gc_prepare_to_collect` doing the actual preparation that is done in `jl_gc_collect` from `gc.c`.
We could potentially reuse `jl_gc_collect` for what I'm doing by passing a different argument (eg. -1), but we'd still have the duplication - the code for Julia in `gc.c` and the code for MMTk in `mmtk-gc.c`.